### PR TITLE
feat(entities): Implement rest of AllayMock

### PR DESCRIPTION
**Pull request by thelooter**
# Description
Implements the duplicate and dancing APIs

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Code follows existing style.
- [x] Unit tests added (if applicable).

### CODE REVIEW

File: src/main/java/be/seeseemelk/mockbukkit/entity/AllayMock.java [hunk 1]
```diff
@@ -30,6 +31,10 @@ public class AllayMock extends CreatureMock implements Allay
 
 	private final @NotNull Inventory inventory;
 	private Material currentItem;
+	private boolean canDulicate = true;
```
Comment (Thorinwasher) [inline src/main/java/be/seeseemelk/mockbukkit/entity/AllayMock.java]: Typo

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java [hunk 2]
```diff
@@ -1229,8 +1230,7 @@ public boolean fromMobSpawner()
 	@Override
 	public CreatureSpawnEvent.@NotNull SpawnReason getEntitySpawnReason()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return this.spawnReason;
```
Comment (Thorinwasher) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java:1233]: Should these changes in this class be part of the pull request? What I understand, they do not relate to AllayMock.

(Personally it's fine for me as is)

File: src/main/java/be/seeseemelk/mockbukkit/entity/AllayMock.java [hunk 3]
```diff
@@ -150,78 +155,79 @@ public void assertInventoryContains(ItemStack item, String s)
...
 	public void startDancing(@NotNull Location location)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.checkArgument(location != null, "Location cannot be null");
+		Preconditions.checkArgument(location.getBlock().getType() == Material.JUKEBOX,
+				"Location must be a Jukebox");
+
+		this.isDancing = true;
+		this.jukebox = location;
```
Comment (Thorinwasher) [inline src/main/java/be/seeseemelk/mockbukkit/entity/AllayMock.java]:
```suggestion
		this.jukebox = location.clone();
```
Some weird behavior could come from using the same reference as provided externally. There's for example the Location#add() method which can modify the object externally after the location has been set.

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java (hunk 2 above)
Comment (thelooter) [reply src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java:1233]: I put it in here, since this was needed to Implement `Allay#duplicateAllay()`

Comment (Thorinwasher) [reply src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java:1233]: Alright

File: src/main/java/be/seeseemelk/mockbukkit/entity/AllayMock.java [hunk 4]
```diff
@@ -150,78 +154,79 @@ public void assertInventoryContains(ItemStack item, String s)
...
+
+		return allay;
 	}
 
 	@Override
 	public @Nullable Location getJukebox()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return this.jukebox;
```
Comment (Thorinwasher) [inline src/main/java/be/seeseemelk/mockbukkit/entity/AllayMock.java]:
```suggestion
		return this.jukebox.clone();
```
You should also do this (same reasoning as in startdancing)

**Review by Thorinwasher — APPROVED** [review]

**Pull request merged**